ERR_PNPM_FETCH_401

Hi,

I have an issue with `pnpm`, when I'm trying to install my private npm packages from my private npm registry (in this case AWS Codeartifact), I'm getting the error:

```
#11 32.53  ERR_PNPM_FETCH_401  GET https://<mydomain>-dom-<accountid>.d.codeartifact.us-east-1.amazonaws.com:443/npm/<myrepo>-repo/@<myorg2>/<mypackage>/-/<mypackage>-1.0.35.tgz: Unauthorized - 401
#11 32.53 
#11 32.53 No authorization header was set for the request.
#11 32.53 
#11 32.53 These authorization settings were found:
#11 32.53 @<myorg1>:registry=https://<mydomain>-dom-<accountid>.d.codeartifact.us-east-1.amazonaws.com/npm/<myrepo>-repo/
#11 32.53 @<myorg2>registry=https://<mydomain>-dom-<accountid>.d.codeartifact.us-east-1.amazonaws.com/npm/<myrepo>-repo/
```

here is my `.npmrc`:
```
registry=https://registry.npmjs.org/
@<myorg1>:registry=https://<mydomain>-dom-<accountid>.d.codeartifact.us-east-1.amazonaws.com/npm/<myrepo>-repo/
@<myorg2>:registry=https://<mydomain>-dom-<accountid>.d.codeartifact.us-east-1.amazonaws.com/npm/<myrepo>-repo/
//<mydomain>-dom-<accountid>.d.codeartifact.us-east-1.amazonaws.com/npm/<myrepo>-repo/:always-auth=true
//<mydomain>-dom-<accountid>.d.codeartifact.us-east-1.amazonaws.com/npm/<myrepo>-repo/:_authToken=<mytoken>
```

package.json:
```
...
"dependencies": {
    "@<myorg2>/<mypackage>": "1.0.35",
}
...
```

 - node:18.14.0-bullseye-slim
 - Linux
 - pnpm version: 8.1.1

--- a/.changeset/quiet-oranges-end.md
+++ b/.changeset/quiet-oranges-end.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/network.auth-header": patch
+"pnpm": patch
+---
+
+Ignore the port in the URL, while searching for authentication token in the `.npmrc` file [#6354](https://github.com/pnpm/pnpm/issues/6354).

--- a/network/auth-header/src/index.ts
+++ b/network/auth-header/src/index.ts
@@ -1,4 +1,5 @@
 import nerfDart from 'nerf-dart'
+import { URL } from 'url'
 import { getAuthHeadersFromConfig } from './getAuthHeadersFromConfig'
 
 export function createGetAuthHeaderByURI (
@@ -22,12 +23,23 @@ function getMaxParts (uris: string[]) {
   }, 0)
 }
 
-function getAuthHeaderByURI (authHeaders: Record<string, string>, maxParts: number, uri: string) {
+function getAuthHeaderByURI (authHeaders: Record<string, string>, maxParts: number, uri: string): string | undefined {
   const nerfed = nerfDart(uri)
   const parts = nerfed.split('/')
   for (let i = Math.min(parts.length, maxParts) - 1; i >= 3; i--) {
     const key = `${parts.slice(0, i).join('/')}/` // eslint-disable-line
     if (authHeaders[key]) return authHeaders[key]
   }
+  const urlWithoutPort = removePort(uri)
+  if (urlWithoutPort !== uri) {
+    return getAuthHeaderByURI(authHeaders, maxParts, urlWithoutPort)
+  }
   return undefined
+}
+
+function removePort (originalUrl: string) {
+  const urlObj = new URL(originalUrl)
+  if (urlObj.port === '') return originalUrl
+  const newUrlObj = new URL(`${urlObj.protocol}//${urlObj.hostname}${urlObj.pathname}${urlObj.search}${urlObj.hash}`)
+  return newUrlObj.toString()
 }

--- a/network/auth-header/test/getAuthHeaderByURI.ts
+++ b/network/auth-header/test/getAuthHeaderByURI.ts
@@ -5,13 +5,16 @@ test('getAuthHeaderByURI()', () => {
     allSettings: {
       '//reg.com/:_authToken': 'abc123',
       '//reg.co/tarballs/:_authToken': 'xxx',
+      '//reg.gg:8888/:_authToken': '0000',
     },
     userSettings: {},
   })
   expect(getAuthHeaderByURI('https://reg.com/')).toBe('Bearer abc123')
   expect(getAuthHeaderByURI('https://reg.com/foo/-/foo-1.0.0.tgz')).toBe('Bearer abc123')
+  expect(getAuthHeaderByURI('https://reg.com:8080/foo/-/foo-1.0.0.tgz')).toBe('Bearer abc123')
   expect(getAuthHeaderByURI('https://reg.io/foo/-/foo-1.0.0.tgz')).toBe(undefined)
   expect(getAuthHeaderByURI('https://reg.co/tarballs/foo/-/foo-1.0.0.tgz')).toBe('Bearer xxx')
+  expect(getAuthHeaderByURI('https://reg.gg:8888/foo/-/foo-1.0.0.tgz')).toBe('Bearer 0000')
 })
 
 test('returns undefined when the auth header is not found', () => {

